src/main/java/com/aws/greengrass/authorization/WildcardVariableTrie.java

```diff
@@ -0,0 +1,169 @@
+        // '*' alone allows all resources including multiple levels
+        if (subject.equals(GLOB_WILDCARD)) {
+            this.matchAll = true;
+            return this;
+        }
+        WildcardVariableTrie current = this;
+        StringBuilder sb = new StringBuilder(subject.length());
+        for (int i = 0; i < subject.length(); i++) {
+            // Maybe create node for a wildcard
+            if (subject.charAt(i) == '*') {
+                current = current.add(sb.toString(), false);
+                current = current.children.get(GLOB_WILDCARD);
+                current.isGlobWildcard = true;
+                // If the string ends with *, then the wildcard is a terminal
+                if (i == subject.length() - 1) {
+                    current.isTerminal = isTerminal;
+                    return current;
+                }
+                return current.add(subject.substring(i + 1), true);
+            }
+            // Create a node for MQTT wildcard, should be terminal and preceded with '/'
+            if (i > 0 && subject.charAt(i) == '#' && subject.charAt(i - 1) == '/' && i == (subject.length() - 1)) {
+                current = current.add(sb.toString(), false);
+                current = current.children.get(MQTT_WILDCARD);
+                current.isTerminal = true;
+                return current;
+            }
+            // Maybe create node for a variable
+            if (subject.charAt(i) == '$') {
+                String variableValue = getVariable(subject.substring(i));
+                if (variableValue != null) {
+                    current = current.add(sb.toString(), false);
+                    current = current.children.get(variableValue);
+                    current.isVariable = true;
+                    return current.add(subject.substring(i + variableValue.length()), true);
+                }
+            }
+            sb.append(subject.charAt(i));
+        }
+        // Handle non-wildcard and non-variable value
+        current = current.children.get(sb.toString());
+        current.isTerminal |= isTerminal;
+        return current;
+    }
+
+    @Nullable
+    private String getVariable(String str) {
+        // Minimum possible variable would be ${a}
+        if (str.length() < 4) {
+            return null;
+        }
+        // Make sure we match the format ${<variable text>}
+        if (str.charAt(0) == '$' && str.charAt(1) == '{') {
+            int variableEndIndex = str.indexOf('}');
+            if (variableEndIndex > 2) {
+                return str.substring(0, variableEndIndex + 1);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get resources for combination of destination, principal and operation.
+     * Also returns resources covered by permissions with * operation/principal.
+     *
+     * @param str string to match
+     * @param variables map of vars
+     *
+     */
+    public boolean matches(String str, Map<String, String> variables) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.